io/include/traccc/io/csv.hpp

```diff
@@ -235,83 +236,103 @@ inline cell_container_types::host read_cells(
     const Acts::GeometryHierarchyMap<digitization_config>* digi_cfg = nullptr,
     unsigned int max_cells = std::numeric_limits<unsigned int>::max()) {
 
-    cell_container_types::host result(&resource);
+    // Create cell counter vector
+    cell_counter_collection_types::host cell_counts(&resource);
+    // Create cell vector
+    cell_collection_types::host allCells(&resource);
 
     unsigned int read_cells = 0;
     csv_cell iocell;
 
+    // Read all cells from input file
     while (creader.read(iocell)) {
-        cell_module module;
 
         cell c = cell{iocell.channel0, iocell.channel1, iocell.value,
                       iocell.timestamp};
 
-        module.module = iocell.geometry_id;
+        // Add current cell to vector
+        allCells.push_back(c);
 
-        const cell_container_types::host::header_vector& headers =
-            result.get_headers();
+        geometry_id id = iocell.geometry_id;
 
-        auto it = std::find(headers.begin(), headers.end(), module);
+        auto rit = std::find_if(
+            cell_counts.rbegin(), cell_counts.rend(),
+            [&id](const cell_counter& cc) { return cc.module == id; });
 
-        if (it == headers.end()) {
+        if (rit == cell_counts.rend()) {
+            // Add module to cell counter vector
+            cell_counts.push_back({id, 1});
+        } else {
+            // Increment number of cells for this module
+            ++(rit->nCells);
+        }
 
-            if (tfmap != nullptr) {
-                if (!tfmap->contains(module.module)) {
-                    throw std::runtime_error(
-                        "Could not find placement for geometry ID " +
-                        std::to_string(module.module));
-                }
-                module.placement = (*tfmap)[module.module];
-            }
+        if (++read_cells >= max_cells) {
+            break;
+        }
+    }
 
-            if (digi_cfg != nullptr) {
-                Acts::GeometryIdentifier id(module.module);
+    // Construct the size vectors.
+    const std::size_t size = cell_counts.size();
+    std::vector<std::size_t> numCells(size);
+    std::transform(cell_counts.begin(), cell_counts.end(), numCells.begin(),
+                   [](const cell_counter& counter) { return counter.nCells; });
```

Comment:
This transformation seems a bit unnecessary, isn't it? You only access single elements from this `numCells` vector. At that point you might as well just get the single numbers from the original `cell_counts` object in all those places.